shared/qcommon/ojk_saved_game.cpp

```diff
@@ -0,0 +1,974 @@
+	const uint32_t chunk_id)
+{
+	if (is_failed_)
+	{
+		return false;
+	}
+
+	if (file_handle_ == 0)
+	{
+		is_failed_ = true;
+		error_message_ = "Not open or created.";
+		return false;
+	}
+
+	io_buffer_offset_ = 0;
+
+	const auto chunk_id_string = get_chunk_id_string(
+		chunk_id);
+
+	::Com_DPrintf(
+		"Attempting read of chunk %s\n",
+		chunk_id_string.c_str());
+
+	uint32_t ulLoadedChid = 0;
+	uint32_t uiLoadedLength = 0;
+
+	auto uiLoaded = ::FS_Read(
+		&ulLoadedChid,
+		static_cast<int>(sizeof(ulLoadedChid)),
+		file_handle_);
+
+	uiLoaded += ::FS_Read(
+		&uiLoadedLength,
+		static_cast<int>(sizeof(uiLoadedLength)),
+		file_handle_);
+
+	const bool bBlockIsCompressed = (static_cast<int32_t>(uiLoadedLength) < 0);
+
+	if (bBlockIsCompressed)
+	{
+		uiLoadedLength = -static_cast<int32_t>(uiLoadedLength);
+	}
+
+	// Make sure we are loading the correct chunk...
+	//
+	if (ulLoadedChid != chunk_id)
+	{
+		is_failed_ = true;
+
+		const auto loaded_chunk_id_string = get_chunk_id_string(
+			ulLoadedChid);
+
+		error_message_ =
+			"Loaded chunk ID (" +
+				loaded_chunk_id_string +
+				") does not match requested chunk ID (" +
+				chunk_id_string +
+				").";
+
+		return false;
+	}
+
+	uint32_t uiLoadedCksum = 0;
+
+#ifdef JK2_MODE
+	// Get checksum...
+	//
+	uiLoaded += ::FS_Read(
+		&uiLoadedCksum,
+		static_cast<int>(sizeof(uiLoadedCksum)),
+		file_handle_);
+#endif // JK2_MODE
+
+	// Load in data and magic number...
+	//
+	uint32_t uiCompressedLength = 0;
+
+	if (bBlockIsCompressed)
+	{
+		uiLoaded += ::FS_Read(
+			&uiCompressedLength,
+			static_cast<int>(sizeof(uiCompressedLength)),
+			file_handle_);
+
+		rle_buffer_.resize(
+			uiCompressedLength);
+
+		uiLoaded += ::FS_Read(
+			rle_buffer_.data(),
+			uiCompressedLength,
+			file_handle_);
+
+		io_buffer_.resize(
+			uiLoadedLength);
+
+		decompress(
+			rle_buffer_,
+			io_buffer_);
+	}
+	else
+	{
+		io_buffer_.resize(
+			uiLoadedLength);
+
+		uiLoaded += ::FS_Read(
+			io_buffer_.data(),
+			uiLoadedLength,
+			file_handle_);
+	}
+
+#ifdef JK2_MODE
+	uint32_t uiLoadedMagic = 0;
+
+	uiLoaded += ::FS_Read(
+		&uiLoadedMagic,
+		static_cast<int>(sizeof(uiLoadedMagic)),
+		file_handle_);
+
+	if (uiLoadedMagic != get_jo_magic_value())
+	{
+		is_failed_ = true;
+
+		error_message_ =
+			"Bad saved game magic for chunk " + chunk_id_string + ".";
+
+		return false;
+	}
+#endif // JK2_MODE
+
+#ifndef JK2_MODE
+	// Get checksum...
+	//
+	uiLoaded += ::FS_Read(
+		&uiLoadedCksum,
+		static_cast<int>(sizeof(uiLoadedCksum)),
+		file_handle_);
+#endif // !JK2_MODE
+
+	// Make sure the checksums match...
+	//
+	const auto uiCksum = ::Com_BlockChecksum(
+		io_buffer_.data(),
+		static_cast<int>(io_buffer_.size()));
+
+	if (uiLoadedCksum != uiCksum)
+	{
+		is_failed_ = true;
+
+		error_message_ =
+			"Failed checksum check for chunk " + chunk_id_string + ".";
+
+		return false;
+	}
+
+	// Make sure we didn't encounter any read errors...
+	if (uiLoaded !=
+		sizeof(ulLoadedChid) +
```

Comment:
Can we move all of this addition into a variable? And then the `#ifdef JK2_MODE` block can be added on without it looking so messy.